### PR TITLE
feat: add client-side Verilog synthesis via YoWASP WASM for offline Tauri desktop

### LINKED ISSUE #1058: Feature: Introduce Client-Side Verilog Synthesis via Web Worker for Tauri Desktop
### Is your feature request related to a problem? Please describe

Yes. As part of my GSoC project (Project 7), I am bringing offline Verilog Synthesis to the Tauri Desktop application. However, implementing this presents two major technical problems:
1. **Main Thread Blocking**: Running the `@yowasp/yosys` WASM engine directly on the main thread is a heavy operation that freezes the application UI during circuit compilation.
2. **Repository Bloat**: The offline synthesis process requires generating custom worker bundles and downloading large `yosys.core*.wasm` binaries (~15MB). Committing these heavy artifacts directly to the repository is a bad version control practice that will bloat the codebase.

### Describe the solution you'd like

I will implement a robust architectural solution for the desktop app:
1. **Web Worker Integration**: Introduce a dedicated Web Worker to handle the Yosys WASM engine off the main thread, keeping the UI perfectly responsive during synthesis.
2. **Automated Build Pipeline**: Set up a clean, zero-friction developer workflow that automatically generates the worker bundles and handles the WASM binaries locally during `npm install` and `npm run dev`, without needing to track any of these heavy artifacts in version control.

### Describe alternatives you've considered

- **Server-Side API**: Relying on the existing Rails backend API. Rejected because the primary goal is to provide a fully offline capability for the desktop application.
- **Forking the whole yosys CI**: Rejected because it creates unnecessary repository bloat and leads to messy PRs.

### Is requested feature related to vue simulator?

Yes

### Additional context

This feature is a core deliverable for **GSoC 2026 Project 7: Client-Side Verilog Synthesis with Tauri**.

### Are you working on this?

Yes

### LINKED ISSUE #1059: refactor: migrate Tauri environment detection from internals to official isTauri API
### Description
Currently, the worker application relies on Tauri internals/window globals to detect whether it is running within the Tauri desktop environment. This approach is an anti-pattern and relies on undocumented internal behavior which could break in future Tauri updates.

This issue tracks the refactoring of environment detection to use the official `@tauri-apps/api/core` `isTauri()` method.

### Motivation
- **Best Practices:** Using official APIs rather than internal window globals ensures long-term stability.
- **Maintainability:** Makes platform-branching logic cleaner and easier to read.
- **GSoC 2026:** This fulfills a deliverable for the Client-Side Verilog Synthesis project to establish robust offline-routing infrastructure.

### Proposed Changes
- Remove legacy Tauri internal checks.
- Implement `isTauri()` for all desktop-specific branching (e.g., routing offline synthesis vs. server synthesis).

### Related PRs
- This refactor will be included and resolved alongside the upcoming Client-Side Verilog Synthesis PR.

### PR DESCRIPTION
**Pull request by Me-Priyank**
Fixes #1058
Fixes #1059

### Describe the changes you have made in this PR -

This PR adds **client-side Verilog synthesis** to the v1 simulator so the Tauri desktop app can synthesize Verilog fully offline (no backend / no YosysJS server call).

**New files**
- `v1/src/simulator/src/synthesis/synthesisWorker.js`, a Web Worker that runs `@yowasp/yosys` (Yosys compiled to WASM) plus `yosys2digitaljs` to turn Verilog source into a DigitalJS circuit graph, off the main thread.
- `v1/src/simulator/src/synthesis/clientSynthesis.js`, spawns the worker (loaded via Vite's native `new URL(..., import.meta.url)` worker pattern), sends the Verilog input, and returns the synthesized circuit.
- `v1/src/simulator/src/synthesis/circuitLayout.js`, dependency-free auto-layout: `computeLayout()` does a topological layered placement (Kahn's algorithm) and `computePortLayout()` positions the subcircuit shell's input/output ports.
- `v1/src/simulator/spec/synthesis.spec.js`, unit tests for the layout logic.

**Modified files**
- `v1/src/simulator/src/Verilog2CV.js`, wires the client-side synthesis path into the existing Verilog-to-CircuitVerse flow using an `isTauri()` check.
- `vite.config.ts`, strips the `#!` shebang from `yosys2digitaljs` so it bundles cleanly, and sets the `application/wasm` MIME type on the dev server. (Vite bundles the worker and emits the WASM/resource assets automatically, so no manual copy step is needed.)
- `vitest.config.ts`, adds the `synthesis` test project.
- `package.json` / `package-lock.json`, add the `@yowasp/yosys` and `yosys2digitaljs` dependencies and the `test:synthesis` script.
- `v1/src/simulator/src/layoutMode.ts`, fixes an operator-precedence bug in the label-offset math (`x + typeof info[1] === "number" ? ...` always took the else branch).
- `v1/src/simulator/src/subcircuit.ts`, label-rendering adjustment for synthesized subcircuit shells.

### Screenshot of build -

Screenshot 2026-06-24 111330

### Screenrecording -

https://github.com/user-attachments/assets/6f182443-d5f9-4bec-8db9-ceaaf189f02a

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [x] No, I wrote all the code myself
- [ ] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [ ] I have reviewed every single line of the AI-generated code
- [ ] I can explain the purpose and logic of each function/component I added
- [ ] I have tested edge cases and understand how the code handles them
- [ ] I have modified the AI output to follow this project's coding standards and conventions

### Explain your implementation approach:

**What problem does this solve?**

Verilog synthesis in the simulator currently depends on a server-side Yosys call, so it does not work in the offline Tauri desktop build. Running Yosys (a large WASM binary) directly on the browser's main thread would block the event loop and freeze the UI while synthesis runs. The WASM binaries are also large, so they should not be committed to the repository.

**Alternative approaches I considered:**

- Running synthesis on the main thread : rejected, since it freezes the UI for the duration of synthesis.
- Shipping a server-side or bundled native Yosys : rejected, since it defeats the goal of fully offline desktop synthesis and adds a backend / native-build dependency.
- I initially bundled the worker with a standalone Rollup config, but now, project's bundler (Vite 8 / Rolldown) handles the worker and its WASM imports natively without any interfarence of esbuild, so I removed the extra config in favour of Vite's built-in worker support.

**Why I chose this implementation:**

I run synthesis in a dedicated Web Worker so the heavy Yosys WASM execution never blocks the main thread and the UI stays responsive. The worker is loaded with Vite's native worker pattern, `new Worker(new URL('./synthesisWorker.js', import.meta.url), { type: 'module' })`, so Vite automatically bundles the worker and emits the WASM/resource assets at build time. This means no separate build step and no large binaries committed to git: a contributor gets everything just by running `npm install` and `npm run tauri dev` for v1.

**Key components:**
- `synthesisWorker.js`: runs Yosys WASM + yosys2digitaljs in a worker.
- `clientSynthesis.js`: spawns the worker (via Vite's native worker URL), handles progress/result/error messages.
- `circuitLayout.js`: auto-places the synthesized circuit using a topological sort so gates flow left-to-right by dependency; `computePortLayout` positions the shell ports.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

## Summary by CodeRabbit

* **New Features**
  * Added circuit auto-layout for device placement and port geometry.
  * Added simulator Verilog/Yosys synthesis with live progress updates, running in the browser or on desktop depending on platform.
* **Bug Fixes**
  * Improved synthesis error handling and user-facing messaging across web and desktop.
  * Fixed subcircuit label vertical alignment at exact boundary positions.
* **Tests**
  * Added dedicated synthesis spec coverage and updated test scripts/config to run synthesis tests separately.

### CODE REVIEW

File: vitest.config.ts [hunk 1]
```diff
@@ -65,8 +67,16 @@ const createProjectConfig = (version: string) => ({
...
+    name: "synthesis",
+    environment: "node" as const,
+    include: synthesisSpecs,
+  },
+};
+
 export default defineConfig({
   test: {
-    projects: [createProjectConfig("src"), createProjectConfig("v1")],
+    projects: [createProjectConfig("src"), createProjectConfig("v1"), synthesisProject],
```
[the remark this answers is not in the document]
Comment (Me-Priyank) [reply vitest.config.ts:80]: @coderabbitai This isn't a regression from this PR. `createProjectConfig("v0")` was never part of the `projects` matrix, `main` is `projects: [createProjectConfig("src"), createProjectConfig("v1")]`, and this PR only appends `synthesisProject`, giving `[..."src", ..."v1", synthesisProject]`. No v0 suite existed to drop.

The `version === "v0" ? "v1/**/*" : "v0/**/*"` branch inside `createProjectConfig` is pre-existing, unchanged code from `main` , vestigial handling for a `v0` project that upstream also never instantiates. If removing that dead ternary is desirable, it'd be a separate cleanup unrelated to this feature.

File: v1/src/simulator/src/Verilog2CV.js [hunk 2]
```diff
@@ -258,17 +339,24 @@ export default function generateVerilogCircuit(
...
+                    error.json()
+                        .then((errorMessage) => {
+                            setVerilogOutput(errorMessage.message, 'error')
+                        })
+                        .catch(() => {
+                            setVerilogOutput('Server returned a non-JSON error response', 'error')
+                        })
+                }
             }
         })
```
[the remark this answers is not in the document]
Comment (Me-Priyank) [reply v1/src/simulator/src/Verilog2CV.js:363]: fixed

File: v1/src/simulator/src/Verilog2CV.js [hunk 3]
```diff
@@ -68,6 +71,37 @@ export function applyVerilogTheme(theme) {
...
+            } else if (type === 'success') {
+                verilogOutputDiv.innerHTML = text
+                verilogOutputDiv.style.color = '#51cf66'
+            } else {
+                verilogOutputDiv.innerHTML = text
+                verilogOutputDiv.style.color = ''
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (Me-Priyank) [reply v1/src/simulator/src/Verilog2CV.js:90]: fixed

File: v1/src/simulator/src/synthesis/circuitLayout.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        inputs: positions(inputs, 0),
+        outputs: positions(outputs, PORT_LAYOUT_WIDTH),
+    };
+}
+
+// Topological layered layout for a DigitalJS circuit
+export function computeLayout(circuitJSON) {
+    if (!circuitJSON.devices || Object.keys(circuitJSON.devices).length === 0) {
+        return {}
+    }
```
[the remark this answers is not in the document]
Comment (Me-Priyank) [reply v1/src/simulator/src/synthesis/circuitLayout.js]: fixed

**Comment by Me-Priyank** [comment]
@Nihal4777 @092vk @ThatDeparted2061 PTAL

File: package.json [hunk 5]
```diff
@@ -30,6 +31,7 @@
     "@tiptap/extension-underline": "^2.0.3",
     "@tiptap/starter-kit": "^2.0.3",
     "@tiptap/vue-3": "^2.0.3",
+    "@yowasp/yosys": "0.59.84-dev.1027",
```
Comment (ThatDeparted2061) [inline package.json:34]: Is pinning to a `-dev` prerelease (`0.59.84-dev.1027`) intentional? Dev builds on npm can get unpublished / aren't guaranteed to stick around, so this could break fresh installs and CI down the line. If there's a stable `@yowasp/yosys` release that works, I'd lean towards that for reproducibility.

File: vite.config.ts [hunk 6]
```diff
@@ -8,20 +8,52 @@ import vuetify from "vite-plugin-vuetify";
...
+      wasmMimePlugin(),
       vue(),
       vuetify({ autoImport: true }),
       cssInjectedByJsPlugin(),
       VueI18nPlugin({
         strictMessage: false,
       }),
     ],
+    optimizeDeps: {
+      exclude: ["@yowasp/yosys", "vue"],
```
Comment (ThatDeparted2061) [inline vite.config.ts:58]: Why is `vue` in the `optimizeDeps.exclude` list here? Excluding `@yowasp/yosys` makes sense for the WASM, but `vue` looks unrelated to the synthesis path and excluding it from prebundling can slow the dev server / change resolution behavior. If it's not actually needed for this feature, probably worth dropping.

File: v1/src/simulator/src/synthesis/circuitLayout.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+            // Center the layer vertically
+            var totalHeight = 0
+            for (var n = 0; n < nodesInLayer.length; n++) {
+                var size = DEVICE_SIZES[circuitJSON.devices[nodesInLayer[n]].type] || DEFAULT_SIZE
+                totalHeight += size.height
+            }
+            totalHeight += (nodesInLayer.length - 1) * NODE_GAP
+
+            var startY = MARGIN_Y + Math.max(0, (400 - totalHeight) / 2)
```
Comment (ThatDeparted2061) [inline v1/src/simulator/src/synthesis/circuitLayout.js]: add a comment for what 400 is

File: v1/src/simulator/src/synthesis/clientSynthesis.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
+// Client-side synthesis — spawns a web worker for YoWASP Yosys
+
+var worker = null
+var requestId = 0
+
+export function synthesizeVerilog(verilogCode, onProgress) {
+    return new Promise((resolve, reject) => {
+        if (!worker) {
+            try {
+                worker = new Worker(new URL('./synthesisWorker.js', import.meta.url), { type: 'module' })
```
Comment (ThatDeparted2061) [inline v1/src/simulator/src/synthesis/clientSynthesis.js:10]: Minor: the worker is created once and kept alive for the lifetime of the page but never `terminate()`d. Reusing it is reasonable (avoids re-loading the WASM each time), just flagging in case you'd want to free it when the Verilog editor is closed.

**Review by ThatDeparted2061 — COMMENTED** [review]
Went through this overall a really nice approach. Keeping the WASM binaries out of git and pushing the heavy Yosys work into a worker so the UI doesn't freeze is exactly the right call, and the layout helpers having actual unit coverage is great to see. The operator-precedence fix in `layoutMode.ts` is a good catch too.

mostly the `@yowasp/yosys` dev-version pin and a couple of minor cleanups. Nothing blocking from my side, just things worth a look before this goes in.

File: v1/src/simulator/src/synthesis/clientSynthesis.js (hunk 8 above)
Comment (Me-Priyank) [reply v1/src/simulator/src/synthesis/clientSynthesis.js:10]: @ThatDeparted2061 Yes, the reuse is intentional, it avoids re-downloading and re-initializing the Yosys WASM on every synthesis. PR #1105 already adds `terminate()` (on timeout, the worker is killed and a fresh one is lazily spawned on the next run). And my next deliverable covers periodic worker respawn, terminating and recreating the worker after N runs to free the Yosys linear-memory heap, which grows across runs. Freeing the worker when the Verilog editor is closed is a good idea too, I'll fold that into that deliverable.

File: package.json (hunk 5 above)
Comment (Me-Priyank) [reply package.json:34]: @ThatDeparted2061  I pinned it on a CodeRabbit suggestion https://github.com/CircuitVerse/cv-frontend-vue/pull/1112#discussion_r3480807845 and it felt right for reproducibility, an exact version means every fresh install and CI run gets the identical build instead of drifting within the caret range. That said, you're right that a `-dev` build can be unpublished, so I'll check whether the latest stable `@yowasp/yosys` works with the synthesis script and switch to it if it does.

File: vite.config.ts (hunk 6 above)
Comment (Me-Priyank) [reply vite.config.ts:58]: @ThatDeparted2061 removing `vue` breaks the dev server with `Uncaught ReferenceError: init_runtime_dom_esm_bundler is not defined`. Vite pre-bundling Vue's esm-bundler runtime alongside vuetify + vue-i18n generates a broken init wrapper. The build passes either way (optimizeDeps is dev-only), which is why it wasn't obvious. I've kept vue excluded and added a comment explaining why. 

Screenshot 2026-06-29 105620

File: v1/src/simulator/src/synthesis/circuitLayout.js (hunk 7 above)
Comment (Me-Priyank) [reply v1/src/simulator/src/synthesis/circuitLayout.js]: @ThatDeparted2061 Added

File: v1/src/simulator/src/synthesis/clientSynthesis.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+                case 'error':
+                    cleanup()
+                    reject(new Error(e.data.message))
+                    break
+            }
+        }
+
+        function handleError(err) {
+            cleanup()
```
Comment (092vk) [inline v1/src/simulator/src/synthesis/clientSynthesis.js:40]: after cleanup set,  worker=null , even after the crash , the worker can contain the ref of crashed worker, causing the if(!worker) condition to fail.

Comment (Me-Priyank) [reply v1/src/simulator/src/synthesis/clientSynthesis.js:40]: @092vk  fixed...

**Comment by Me-Priyank** [comment]
@092vk @ThatDeparted2061 PTAL

**Comment by Me-Priyank** [comment]
Finally 😭. thanks @ThatDeparted2061 @092vk .. feels like a major achievement to me
